xarray/core/accessors.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from .common import is_datetime_like
+from .extensions import register_dataarray_accessor
+
+import pandas as pd
+
+def _get_date_field(array, name):
+    """Indirectly access pandas' libts.get_date_field by wrapping data
+    as a Series and calling through `.dt` attribute.
+
+    Parameters
+    ----------
+    array : np.ndarray or array-like
+        Array-like container of datetime-like values
+    name : str
+        Name of datetime field to access
+
+    """
+    series = pd.Series(array.ravel())
+    field_values = getattr(series.dt, name).values
+    return field_values.reshape(array.shape)
+
+_date_field_ops_and_descrs = [
+    ('year', "The year of the datetime"),
+    ('month', "The month as January=1, December=12"),
+    ('day', "The days of the datetime"),
+    ('hour', "The hours of the datetime"),
+    ('minute', "The minutes of the datetime"),
+    ('second', "The seconds of the datetime"),
+    ('microsecond', "The microseconds of the datetime"),
+    ('nanosecond', "The nanoseconds of the datetime"),
+    ('weekofyear',  "The week ordinal of the year"),
+    ('week', "The week ordinal of the year"),
+    ('weekday', "The day of the week with Monday=0, Sunday=6"),
+    ('weekday_name',
+     "The name of day in a week (ex: Friday)"),
+    ('dayofweek', "The day of the week with Monday=0, Sunday=6"),
+    ('dayofyear',  "The ordinal day of the year"),
+    ('quarter', "The quarter of the date"),
+    ('days_in_month', "The number of days in the month"),
+    ('daysinmonth', "The number of days in the month"),
+]
+
+def inject_date_field_accessors(cls):
```

Comment:
This is fine, but I actually did like what you did before -- and we'll probably be switching xarray to use that style in the future, instead of injecting things with a loop. Using a function to define the method directly on the class is much more explicit.